Source/Core/Common/Crypto/AES.cpp

```diff
@@ -1,35 +1,420 @@
+  inline void CryptBlock(__m128i* iv, const u8* buf_in, u8* buf_out) const
+  {
+    __m128i block = _mm_loadu_si128((const __m128i*)buf_in);
+
+    if constexpr (AesMode == Mode::Encrypt)
+    {
+      block = _mm_xor_si128(_mm_xor_si128(block, *iv), round_keys[0]);
+
+      for (size_t i = 1; i < Nr; ++i)
+        block = _mm_aesenc_si128(block, round_keys[i]);
+      block = _mm_aesenclast_si128(block, round_keys[Nr]);
+
+      *iv = block;
+    }
+    else
+    {
+      __m128i iv_next = block;
+
+      block = _mm_xor_si128(block, round_keys[0]);
+
+      for (size_t i = 1; i < Nr; ++i)
+        block = _mm_aesdec_si128(block, round_keys[i]);
+      block = _mm_aesdeclast_si128(block, round_keys[Nr]);
+
+      block = _mm_xor_si128(block, *iv);
+      *iv = iv_next;
+    }
+
+    _mm_storeu_si128((__m128i*)buf_out, block);
+  }
+
+  // Takes advantage of instruction pipelining to parallelize.
+  template <size_t NumBlocks>
+  ATTRIBUTE_TARGET("aes")
+  inline void DecryptPipelined(__m128i* iv, const u8* buf_in, u8* buf_out) const
+  {
+    constexpr size_t Depth = NumBlocks;
+
+    __m128i block[Depth];
+    for (size_t d = 0; d < Depth; d++)
+      block[d] = _mm_loadu_si128(&((const __m128i*)buf_in)[d]);
+
+    __m128i iv_next[1 + Depth];
+    iv_next[0] = *iv;
+    for (size_t d = 0; d < Depth; d++)
+      iv_next[1 + d] = block[d];
+
+    for (size_t d = 0; d < Depth; d++)
+      block[d] = _mm_xor_si128(block[d], round_keys[0]);
+
+    // The main speedup is here
+    for (size_t i = 1; i < Nr; ++i)
+      for (size_t d = 0; d < Depth; d++)
+        block[d] = _mm_aesdec_si128(block[d], round_keys[i]);
+    for (size_t d = 0; d < Depth; d++)
+      block[d] = _mm_aesdeclast_si128(block[d], round_keys[Nr]);
+
+    for (size_t d = 0; d < Depth; d++)
+      block[d] = _mm_xor_si128(block[d], iv_next[d]);
+    *iv = iv_next[1 + Depth - 1];
+
+    for (size_t d = 0; d < Depth; d++)
+      _mm_storeu_si128(&((__m128i*)buf_out)[d], block[d]);
+  }
+
+  virtual bool Crypt(const u8* iv, u8* iv_out, const u8* buf_in, u8* buf_out,
+                     size_t len) const override
+  {
+    if (len % BLOCK_SIZE)
+      return false;
+
+    __m128i iv_block = iv ? _mm_loadu_si128((const __m128i*)iv) : _mm_setzero_si128();
+
+    if constexpr (AesMode == Mode::Decrypt)
+    {
+      // On amd zen2...(benchmark, not real-world):
+      // With AES(128) instructions, BLOCK_DEPTH results in following speedup vs. non-pipelined: 4:
+      // 18%, 8: 22%, 9: 26%, 10-15: 31%. 16: 8% (register exhaustion). With VAES(VEX.128), 10 gives
+      // 36% speedup vs. its corresponding baseline. VAES(VEX.128) is ~4% faster than AES(128). The
+      // result is similar on zen3.
+      // Zen3 in general is 20% faster than zen2 in aes, and VAES(VEX.256) is 35% faster than
+      // zen3/VAES(VEX.128).
+      //  It seems like VAES(VEX.256) should be faster?
+      // TODO Choose value at runtime based on some criteria?
+      constexpr size_t BLOCK_DEPTH = 10;
```

Comment:
btw this is a value that may have a better default, as I've only tested on zen2+3. Maybe someone can try some speed tests  when tweaking it on zen1 or some intel archs?